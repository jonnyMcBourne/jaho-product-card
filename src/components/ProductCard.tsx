import React from "react";
import { createContext, CSSProperties } from "react";
import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import { ProductContextProps, Product, onChangeArgs,InitialValues, ProductCardHandle } from "../interfaces/productInterfaces";

export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: (args:ProductCardHandle) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?:(arg:onChangeArgs)=>void;
  value?:number;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const { counter, increaseBy, isMaxCountReached, maxCount, resetCounter } =
    useProduct({ product, onChange, value, initialValues });

  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          increaseBy: increaseBy,
          isMaxCountReached: isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product: product,
          reset: resetCounter,
        })}
      </div>
    </Provider>
  );
};
export default ProductCard;


