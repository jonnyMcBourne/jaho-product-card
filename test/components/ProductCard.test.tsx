import React from "react";
import renderer from 'react-test-renderer';
import { ProductCard,ProductButtons } from '../../src/components';
import { products } from '../data/products';

describe('ProductCard', () => {
  test('must show the component correctly', () => {
    const wrapper = renderer.create(
      <ProductCard product={products[1]}>
        {({}) => (
          <>
            <ProductButtons />
          </>
        )}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('must increase in 1 the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={products[1]}>
        {({count, increaseBy}) =>{
            return (
                <>
                  <h1> { count } </h1>
                  <button onClick={()=>{increaseBy(1)} }>+1</button>
                </>
              )
        } }
      </ProductCard> 
    );
    
    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(()=>{
        (tree as any).children[1].props.onClick();
    })
    tree = wrapper.toJSON();
    
    expect( (tree as any).children[0].children[1] ).toBe("1");
    });  
});