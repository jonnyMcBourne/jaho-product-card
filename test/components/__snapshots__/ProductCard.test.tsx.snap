// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard must increase in 1 the counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
     
    0
     
  </h1>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;

exports[`ProductCard must show the component correctly 1`] = `
<div
  className="productCard undefined"
>
  <div
    className="buttonsContainer undefined"
  >
    <button
      className="buttonMinus"
      onClick={[Function]}
    >
      -
    </button>
    <div
      className="countLabel"
    >
      0
    </div>
    <button
      className="buttonAdd false"
      disabled={false}
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;
